nano/node/bootstrap/message_deserializer.cpp

```diff
@@ -0,0 +1,350 @@
+#include <nano/node/bootstrap/message_deserializer.hpp>
+#include <nano/node/node.hpp>
+
+nano::bootstrap::message_deserializer::message_deserializer (nano::network_constants const & network_constants, nano::network_filter & publish_filter, nano::block_uniquer & block_uniquer, nano::vote_uniquer & vote_uniquer) :
+	read_buffer{ std::make_shared<std::vector<uint8_t>> () },
+	network_constants{ network_constants },
+	publish_filter{ publish_filter },
+	block_uniquer{ block_uniquer },
+	vote_uniquer{ vote_uniquer }
+{
+	read_buffer->resize (MAX_MESSAGE_SIZE);
+}
+
+void nano::bootstrap::message_deserializer::read (std::shared_ptr<nano::socket> socket, const nano::bootstrap::message_deserializer::callback_type && callback)
+{
+	debug_assert (callback);
+	socket->async_read (read_buffer, HEADER_SIZE, [this_l = shared_from_this (), socket, callback = std::move (callback)] (boost::system::error_code const & ec, std::size_t size_a) {
+		if (ec)
+		{
+			callback (ec, nullptr);
+			return;
+		}
+		if (size_a != HEADER_SIZE)
+		{
+			callback (boost::asio::error::fault, nullptr);
+			return;
+		}
+		this_l->received_header (socket, std::move (callback));
+	});
+}
+
+void nano::bootstrap::message_deserializer::received_header (std::shared_ptr<nano::socket> socket, const nano::bootstrap::message_deserializer::callback_type && callback)
+{
+	nano::bufferstream stream{ read_buffer->data (), HEADER_SIZE };
+	auto error = false;
+	nano::message_header header{ error, stream };
+	if (error)
+	{
+		status = parse_status::invalid_header;
+		callback (boost::system::error_code{}, nullptr);
```

Comment:
These are early-return error checks but passing a non-error code in to the callback.